doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -476,14 +504,138 @@ In Ltac2, we have the following backtracking primitives, defined in the
   val plus : (unit -> 'a) -> (exn -> 'a) -> 'a
   val case : (unit -> 'a) -> ('a * (exn -> 'a)) result
 
-If one views thunks as lazy lists, then `zero` is the empty list and `plus` is
-list concatenation, while `case` is pattern-matching.
+`zero` takes an exception and raises it. It can be understood as the
+trivial exception handler which performs no handling.
+
+`plus` takes a computation `r` and an exception handler `h`, and runs
+the computation `r` under the handler `h`. If `r ()` returns a value
+`a`, `plus` returns `a`; if we write `h'` for the continuation
+implicitly returned by `r ()`, then the continuation associated to the
+expression `plus r h` is behaviorally equivalent to `fun e => plus
+(h' e) h`.
+Else, if `r ()` raises an exception `e`, then `plus r h` is
+behaviorally equivalent to `h e`.
+
+`case` is a "safe" exception handler which always returns a value and
+never re-raises an exception. If `r ()` would raise an exception `e`,
+then `case r` returns `Err e`; otherwise, if `r ()` is equal to a
+value `t`, then `case r = Val(t, f)`, where `f` is the
+exception-handling continuation *implicitly* returned by `r ()`. Thus,
+`case` lets us access directly as values the exceptions and
+backtracking continuations that normally live in the effects monad.
+
+By repeated application of `plus`, it is possible to stack exception
+handlers. If `h_1, h_2 : exn -> 'a` are exception handlers, then `fun
+e => plus (fun () => h1 e ()) h2` is an exception handler which first
+attempts to handle the exception `e` using `h_1`, and, if this fails
+and raises an exception `e'`, attempts to handle `e'` with `h_2`.
+
+  .. example::
+
+     .. coqtop:: in
+
+        (** Convert a list of thunked terms into a single thunked term that behaves as \
+        a stream of values from the original list. Whenever a downstream computation \
+        fails, the next value from the stream is returned. *)
+        Ltac2 rec to_stream (s : (unit -> constr) list) : unit -> constr :=
+          match s with
+          | [] => (fun () => Control.zero No_value)
+          | hd :: tail => (fun () => Control.plus hd (fun _ => to_stream tail ()))
+          end.
+
+        Goal True.
+          let v := List.map (fun a => (fun () => a)) [ '(0%nat); '(1%nat); '2%nat ] in
+          let n := to_stream v () in
+          try (Message.print (Message.of_constr n); fail).
+          (** Prints 0, 1, 2. *)
+
+     .. coqtop:: all abort
+
+This illustrates that one can view thunks as lazy lists of results,
```

Comment:
@ppedrot I kept your original phrasing here, accompanied by an example which makes it concrete.